Tidy Command Line/install/tidy.1
.\" tidy man page for the HTML5 fork of Tidy
.TH TIDY 1 "2014/11/22" "HTML Tidy" "2014/11/22"

.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH NAME
\fBtidy\fR - check, correct, and pretty-print HTML(5) files
.SH SYNOPSIS
\fBtidy\fR [option ...] [file ...] [option ...] [file ...]
.SH DESCRIPTION
Tidy reads HTML(5), XHTML(5) and XML files and writes cleaned-up markup.  For HTML variants, it detects, reports, and corrects many common coding errors and strives to produce visually equivalent markup that is both conformant to the HTML specifications and that works in most browsers.
.LP
A common use of Tidy is to convert plain HTML to XHTML.  For generic XML files, Tidy is limited to correcting basic well-formedness errors and pretty printing.
.LP
If no input file is specified, Tidy reads the standard input.  If no output file is specified, Tidy writes the tidied markup to the standard output.  If no error file is specified, Tidy writes messages to the standard error.
For command line options that expect a numerical argument, a default is assumed if no meaningful value can be found. 
.SH OPTIONS

.SS File manipulation
  
.TP
\fB-output <file>\fR, \fB-o <file>\fR
write output to the specified <file>
(\fIoutput-file: <file>\fR)
.TP
\fB-config <file>\fR
set configuration options from the specified <file>

.TP
\fB-file <file>\fR, \fB-f <file>\fR
write errors and warnings to the specified <file>
(\fIerror-file: <file>\fR)
.TP
\fB-modify\fR, \fB-m\fR
modify the original input files
(\fIwrite-back: yes\fR)
.SS Processing directives
  
.TP
\fB-indent\fR, \fB-i\fR
indent element content
(\fIindent: auto\fR)
.TP
\fB-wrap <column>\fR, \fB-w <column>\fR
wrap text at the specified <column>. 0 is assumed if <column> is missing. When this option is omitted, the default of the configuration option "wrap" applies.
(\fIwrap: <column>\fR)
.TP
\fB-upper\fR, \fB-u\fR
force tags to upper case
(\fIuppercase-tags: yes\fR)
.TP
\fB-clean\fR, \fB-c\fR
replace FONT, NOBR and CENTER tags by CSS
(\fIclean: yes\fR)
.TP
\fB-bare\fR, \fB-b\fR
strip out smart quotes and em dashes, etc.
(\fIbare: yes\fR)
.TP
\fB-gdoc\fR, \fB-g\fR
produce clean version of html exported by google docs
(\fIgdoc: yes\fR)
.TP
\fB-numeric\fR, \fB-n\fR
output numeric rather than named entities
(\fInumeric-entities: yes\fR)
.TP
\fB-errors\fR, \fB-e\fR
show only errors and warnings
(\fImarkup: no\fR)
.TP
\fB-quiet\fR, \fB-q\fR
suppress nonessential output
(\fIquiet: yes\fR)
.TP
\fB-omit\fR
omit optional start tags and end tags
(\fIomit-optional-tags: yes\fR)
.TP
\fB-xml\fR
specify the input is well formed XML
(\fIinput-xml: yes\fR)
.TP
\fB-asxml\fR, \fB-asxhtml\fR
convert HTML to well formed XHTML
(\fIoutput-xhtml: yes\fR)
.TP
\fB-ashtml\fR
force XHTML to well formed HTML
(\fIoutput-html: yes\fR)
.TP
\fB-access <level>\fR
do additional accessibility checks (<level> = 0, 1, 2, 3). 0 is assumed if <level> is missing.
(\fIaccessibility-check: <level>\fR)
.SS Character encodings
  
.TP
\fB-raw\fR
output values above 127 without conversion to entities

.TP
\fB-ascii\fR
use ISO-8859-1 for input, US-ASCII for output

.TP
\fB-latin0\fR
use ISO-8859-15 for input, US-ASCII for output

.TP
\fB-latin1\fR
use ISO-8859-1 for both input and output

.TP
\fB-iso2022\fR
use ISO-2022 for both input and output

.TP
\fB-utf8\fR
use UTF-8 for both input and output

.TP
\fB-mac\fR
use MacRoman for input, US-ASCII for output

.TP
\fB-win1252\fR
use Windows-1252 for input, US-ASCII for output

.TP
\fB-ibm858\fR
use IBM-858 (CP850+Euro) for input, US-ASCII for output

.TP
\fB-utf16le\fR
use UTF-16LE for both input and output

.TP
\fB-utf16be\fR
use UTF-16BE for both input and output

.TP
\fB-utf16\fR
use UTF-16 for both input and output

.TP
\fB-big5\fR
use Big5 for both input and output

.TP
\fB-shiftjis\fR
use Shift_JIS for both input and output

.TP
\fB-language <lang>\fR
set the two-letter language code <lang> (for future use)
(\fIlanguage: <lang>\fR)
.SS Miscellaneous
  
.TP
\fB-version\fR, \fB-v\fR
show the version of Tidy

.TP
\fB-help\fR, \fB-h\fR, \fB-?\fR
list the command line options

.TP
\fB-xml-help\fR
list the command line options in XML format

.TP
\fB-help-config\fR
list all configuration options

.TP
\fB-xml-config\fR
list all configuration options in XML format

.TP
\fB-show-config\fR
list the current configuration settings

.SH USAGE
.LP
Use \fB--\fR\fIoptionX valueX\fR for the detailed configuration option "optionX" with argument "valueX".  See also below under \fBDetailed Configuration Options\fR as to how to conveniently group all such options in a single config file.
.LP
Input/Output default to stdin/stdout respectively. Single letter options apart from \fB-f\fR and \fB-o\fR may be combined as in:
.LP
.in 1i
\fBtidy -f errs.txt -imu foo.html\fR
.SH ENVIRONMENT
.TP
.B HTML_TIDY
Name of the default configuration file.  This should be an absolute path, since you will probably invoke \fBtidy\fR from different directories.  The value of HTML_TIDY will be parsed after the compiled-in default (defined with -DTIDY_CONFIG_FILE), but before any of the files specified using \fB-config\fR.
.SH "EXIT STATUS"
.IP 0
All input files were processed successfully.
.IP 1
There were warnings.
.IP 2
There were errors.

.SH ______________________________
.SH "  "
.SH "DETAILED CONFIGURATION OPTIONS"
This section describes the Detailed (i.e., "expanded") Options, which may be specified by preceding each option with \fB--\fR at the command line, followed by its desired value, OR by placing the options and values in a configuration file, and telling tidy to read that file with the \fB-config\fR standard option.
.SH SYNOPSIS
\fBtidy --\fR\fIoption1 \fRvalue1 \fB--\fIoption2 \fRvalue2 [standard options ...]
.br
\fBtidy -config \fIconfig-file \fR[standard options ...]
.SH WARNING
The options detailed here do not include the "standard" command-line options (i.e., those preceded by a single '\fB-\fR') described above in the first section of this man page.
.SH DESCRIPTION
A list of options for configuring the behavior of Tidy, which can be passed either on the command line, or specified in a configuration file.
.LP
A Tidy configuration file is simply a text file, where each option
is listed on a separate line in the form
.LP
.in 1i
\fBoption1\fR: \fIvalue1\fR
.br
\fBoption2\fR: \fIvalue2\fR
.br
etc.
.LP
The permissible values for a given option depend on the option's \fBType\fR.  There are five types: \fIBoolean\fR, \fIAutoBool\fR, \fIDocType\fR, \fIEnum\fR, and \fIString\fR. Boolean types allow any of \fIyes/no, y/n, true/false, t/f, 1/0\fR.  AutoBools allow \fIauto\fR in addition to the values allowed by Booleans.  Integer types take non-negative integers.  String types generally have no defaults, and you should provide them in non-quoted form (unless you wish the output to contain the literal quotes).
.LP
Enum, Encoding, and DocType "types" have a fixed repertoire of items; consult the \fIExample\fR[s] provided below for the option[s] in question.
.LP
You only need to provide options and values for those whose defaults you wish to override, although you may wish to include some already-defaulted options and values for the sake of documentation and explicitness.
.LP
Here is a sample config file, with at least one example of each of the five Types:
.LP
\fI
    // sample Tidy configuration options
    output-xhtml: yes
    add-xml-decl: no
    doctype: strict
    char-encoding: ascii
    indent: auto
    wrap: 76
    repeated-attributes: keep-last
    error-file: errs.txt
\fR
.LP
Below is a summary and brief description of each of the options. They are listed alphabetically within each category.  There are five categories: \fIHTML, XHTML, XML\fR options, \fIDiagnostics\fR options, \fIPretty Print\fR options, \fICharacter Encoding\fR options, and \fIMiscellaneous\fR options.
.LP
.SH OPTIONS

.SS HTML, XHTML, XML options:
  
.TP
\fBadd-xml-decl\fR

Type:    \fIBoolean\fR
.br
Default: \fIno\fR
.br
Example: \fIy/n, yes/no, t/f, true/false, 1/0\fR

This option specifies if Tidy should add the XML declaration when outputting XML or XHTML. Note that if the input already includes an <?xml ... ?> declaration then this option will be ignored. If the encoding for the output is different from "ascii", one of the utf encodings or "raw", the declaration is always added as required by the XML standard.


.rj 1
\fBSee also\fR: \fIchar-encoding\fR, \fIoutput-encoding\fR
.TP
\fBadd-xml-space\fR

Type:    \fIBoolean\fR
.br
Default: \fIno\fR
.br
Example: \fIy/n, yes/no, t/f, true/false, 1/0\fR

This option specifies if Tidy should add xml:space="preserve" to elements such as <PRE>, <STYLE> and <SCRIPT> when generating XML. This is needed if the whitespace in such elements is to be parsed appropriately without having access to the DTD.
.TP
\fBalt-text\fR

Type:    \fIString\fR
.br
Default: \fI-\fR
.br
Default: \fI-\fR

This option specifies the default "alt=" text Tidy uses for <IMG> attributes. This feature is dangerous as it suppresses further accessibility warnings. You are responsible for making your documents accessible to people who can not see the images!
.TP
\fBanchor-as-name\fR

Type:    \fIBoolean\fR
.br
Default: \fIyes\fR
.br
Example: \fIy/n, yes/no, t/f, true/false, 1/0\fR

This option controls the deletion or addition of the name attribute in elements where it can serve as anchor. If set to "yes", a name attribute, if not already existing, is added along an existing id attribute if the DTD allows it. If set to "no", any existing name attribute is removed if an id attribute exists or has been added.
.TP
\fBassume-xml-procins\fR

Type:    \fIBoolean\fR
.br
Default: \fIno\fR
.br
Example: \fIy/n, yes/no, t/f, true/false, 1/0\fR

This option specifies if Tidy should change the parsing of processing instructions to require ?> as the terminator rather than >. This option is automatically set if the input is in XML.
.TP
\fBbare\fR

Type:    \fIBoolean\fR
.br
Default: \fIno\fR
.br
Example: \fIy/n, yes/no, t/f, true/false, 1/0\fR

This option specifies if Tidy should strip Microsoft specific HTML from Word 2000 documents, and output spaces rather than non-breaking spaces where they exist in the input.
.TP
\fBclean\fR

Type:    \fIBoolean\fR
.br
Default: \fIno\fR
.br
Example: \fIy/n, yes/no, t/f, true/false, 1/0\fR

This option specifies if Tidy should strip out surplus presentational tags and attributes replacing them by style rules and structural markup as appropriate. It works well on the HTML saved by Microsoft Office products.


.rj 1
\fBSee also\fR: \fIdrop-font-tags\fR
.TP
\fBcoerce-endtags\fR

Type:    \fIBoolean\fR
.br
Default: \fIyes\fR
.br
Example: \fIy/n, yes/no, t/f, true/false, 1/0\fR

This option specifies if Tidy should coerce a start tag into an end tag in cases where it looks like an end tag was probably intended; for example, given <span>foo <b>bar<b> baz</span>, Tidy will output <span>foo <b>bar</b> baz</span>.
.TP
\fBcss-prefix\fR

Type:    \fIString\fR
.br
Default: \fI-\fR
.br
Default: \fI-\fR


tidylib-src/console/tidy.c
/*
  cmmdline.c - HTML Tidy command line driver

  Copyright (c) 1998-2003 World Wide Web Consortium
  (Massachusetts Institute of Technology, Institut National de
  Recherche en Informatique et en Automatique, Keio University).
  All Rights Reserved.

  CVS Info :

    $Author: creitzel $ 
    $Date: 2003/02/16 19:33:04 $ 
    $Revision: 1.2 $ 
*/

#include "tidy.h"

uint  contentErrors = 0;
uint  contentWarnings = 0;
uint  optionErrors = 0;
uint  accessWarnings = 0;

FILE* errout = null;  /* set to stderr */
FILE* txtout = null;  /* set to stdout */

Bool samefile( ctmbstr filename1, ctmbstr filename2 )
{
#if FILENAMES_CASE_SENSITIVE
    return ( strcmp( filename1, filename2 ) == 0 );
#else
    return ( strcasecmp( filename1, filename2 ) == 0 );
#endif
}

void help( TidyDoc tdoc, ctmbstr prog )
{
    printf( "%s [option...] [file...] [option...] [file...]\n", prog );
    printf( "Utility to clean up and pretty print HTML/XHTML/XML\n");
    printf( "see http://tidy.sourgeforge.net/\n");
    printf( "\n");

#ifdef PLATFORM_NAME
    printf( "Options for HTML Tidy for %s released on %s:\n",
             PLATFORM_NAME, tidyReleaseDate() );
#else
    printf( "Options for HTML Tidy released on %s:\n", tidyReleaseDate() );
#endif
    printf( "\n");

    printf( "File manipulation\n");
    printf( "-----------------\n");
    printf( "  -out or -o <file> specify the output markup file\n");
    printf( "  -config <file>    set configuration options from the specified <file>\n");
    printf( "  -f      <file>    write errors to the specified <file>\n");
    printf( "  -modify or -m     modify the original input files\n");
    printf( "\n");

    printf( "Processing directives\n");
    printf( "---------------------\n");
    printf( "  -indent  or -i    indent element content\n");
    printf( "  -wrap <column>    wrap text at the specified <column> (default is 68)\n");
    printf( "  -upper   or -u    force tags to upper case (default is lower case)\n");
    printf( "  -clean   or -c    replace FONT, NOBR and CENTER tags by CSS\n");
    printf( "  -bare    or -b    strip out smart quotes and em dashes, etc.\n");
    printf( "  -numeric or -n    output numeric rather than named entities\n");
    printf( "  -errors  or -e    only show errors\n");
    printf( "  -quiet   or -q    suppress nonessential output\n");
    printf( "  -omit             omit optional end tags\n");
    printf( "  -xml              specify the input is well formed XML\n");
    printf( "  -asxml            convert HTML to well formed XHTML\n");
    printf( "  -asxhtml          convert HTML to well formed XHTML\n");
    printf( "  -ashtml           force XHTML to well formed HTML\n");

/* TRT */
#if SUPPORT_ACCESSIBILITY_CHECKS
    printf( "  -access <level>   do additional accessibility checks (<level> = 1, 2, 3)\n");
#endif

    printf( "\n");

    printf( "Character encodings\n");
    printf( "-------------------\n");
    printf( "  -raw              output values above 127 without conversion to entities\n");
    printf( "  -ascii            use US-ASCII for output, ISO-8859-1 for input\n");
    printf( "  -latin0           use US-ASCII for output, ISO-8859-1 for input\n");
    printf( "  -latin1           use ISO-8859-1 for both input and output\n");
    printf( "  -iso2022          use ISO-2022 for both input and output\n");
    printf( "  -utf8             use UTF-8 for both input and output\n");
    printf( "  -mac              use MacRoman for input, US-ASCII for output\n");
    printf( "  -win1252          use Windows-1252 for input, US-ASCII for output\n");
    printf( "  -ibm858           use IBM-858 (CP850+Euro) for input, US-ASCII for output\n");

#if SUPPORT_UTF16_ENCODINGS
    printf( "  -utf16le          use UTF-16LE for both input and output\n");
    printf( "  -utf16be          use UTF-16BE for both input and output\n");
    printf( "  -utf16            use UTF-16 for both input and output\n");
#endif

#if SUPPORT_ASIAN_ENCODINGS /* #431953 - RJ */
    printf( "  -big5             use Big5 for both input and output\n"); 
    printf( "  -shiftjis         use Shift_JIS for both input and output\n");
    printf( "  -language <lang>  set the two-letter language code <lang> (for future use)\n");
#endif
    printf( "\n");

    printf( "Miscellaneous\n");
    printf( "-------------\n");
    printf( "  -version  or -v   show the version of Tidy\n");
    printf( "  -help, -h or -?   list the command line options\n");
    printf( "  -help-config      list all configuration options\n");
    printf( "  -show-config      list the current configuration settings\n");
    printf( "\n");
    printf( "Use --blah blarg for any configuration option \"blah\" with argument \"blarg\"\n");
    printf( "\n");

    printf( "Input/Output default to stdin/stdout respectively\n");
    printf( "Single letter options apart from -f may be combined\n");
    printf( "as in:  tidy -f errs.txt -imu foo.html\n");
    printf( "For further info on HTML see http://www.w3.org/MarkUp\n");
    printf( "\n");
}

#define kMaxValFieldWidth 40
static const char* fmt = "%-27.27s %-9.9s  %-40.40s\n";
static const char* valfmt = "%-27.27s %-9.9s %-1.1s%-39.39s\n";
static const char* ul 
        = "=================================================================";

void optionhelp( TidyDoc tdoc, ctmbstr prog )
{
    TidyIterator pos = tidyGetOptionList( tdoc );

    printf( "\nHTML Tidy Configuration Settings\n\n" );
    printf( "Within a file, use the form:\n\n" ); 
    printf( "wrap: 72\n" );
    printf( "split: no\n\n" );
    printf( "When specified on the command line, use the form:\n\n" );
    printf( "--wrap 72 --split no\n\n");

    printf( fmt, "Name", "Type", "Allowable values" );
    printf( fmt, ul, ul, ul );

    while ( pos )
    {
        TidyOption topt = tidyGetNextOption( tdoc, &pos );
        TidyOptionId optId = tidyOptGetId( topt );
        TidyOptionType optTyp = tidyOptGetType( topt );

        tmbstr name = (tmbstr) tidyOptGetName( topt );
        tmbstr type = "String";
        tmbchar tempvals[80] = {0};
        tmbstr vals = &tempvals[0];

        if ( tidyOptIsReadOnly(topt) )
            continue;

        /* Handle special cases first.
        */
        switch ( optId )
        {
        case TidyIndentContent:
#if SUPPORT_UTF16_ENCODINGS
        case TidyOutputBOM:
#endif
            type = "AutoBool";
            vals = "auto, y/n, yes/no, t/f, true/false, 1/0";
            break;

        case TidyDuplicateAttrs:
            type = "enum";
            vals = "keep-first, keep-last";
            break;

        case TidyDoctype:
            type = "DocType";
            vals = "auto, omit, strict, loose, transitional,";
            printf( fmt, name, type, vals );
            name = "";
            type = "";
            vals = "user specified fpi (string)";
            break;

        case TidyCSSPrefix:
            type = "Name";
            vals = "CSS1 selector";
            break;

        case TidyInlineTags:
        case TidyBlockTags:
        case TidyEmptyTags:
        case TidyPreTags:
            type = "Tag names";
            vals = "tagX, tagY, ...";
            break;

        case TidyCharEncoding:
        case TidyInCharEncoding:
        case TidyOutCharEncoding:
            type = "Encoding";
            vals = "ascii, latin0, latin1, raw, utf8, iso2022,";
            printf( fmt, name, type, vals );
            name = "";
            type = "";

#if SUPPORT_UTF16_ENCODINGS
            vals = "utf16le, utf16be, utf16,";
            printf( fmt, name, type, vals );
#endif
#if SUPPORT_ASIAN_ENCODINGS
            vals = "mac, win1252, ibm858, big5, shiftjis";
#else
            vals = "mac, win1252, ibm858";
#endif
            break;

        case TidyNewline:
            type = "enum";
            vals = "LF, CRLF, CR";
            break;

        /* General case will handle remaining */
        default:
            switch ( optTyp )
            {
            case TidyBoolean:
                type = "Boolean";
                vals = "y/n, yes/no, t/f, true/false, 1/0";
                break;

            case TidyInteger:
                type = "Integer";
                if ( optId == TidyWrapLen )
                    vals = "0 (no wrapping), 1, 2, ...";
                else
                    vals = "0, 1, 2, ...";
                break;

            case TidyString:
                type = "String";
                vals = "-";
                break;
            }
        }

        if ( *name || *type || *vals )
            printf( fmt, name, type, vals );
    }
}

void optionvalues( TidyDoc tdoc, ctmbstr prog )
{
    TidyIterator pos = tidyGetOptionList( tdoc );

    printf( "\nConfiguration File Settings:\n\n" );
    printf( fmt, "Name", "Type", "Current Value" );
    printf( fmt, ul, ul, ul );

    while ( pos )
    {
        TidyOption topt = tidyGetNextOption( tdoc, &pos );
        TidyOptionId optId = tidyOptGetId( topt );
        TidyOptionType optTyp = tidyOptGetType( topt );
        Bool isReadOnly = tidyOptIsReadOnly( topt );

        Bool bval = no;
        ctmbstr sval = null;
        uint ival = 0;

        tmbstr name = (tmbstr) tidyOptGetName( topt );
        tmbstr type = "String";
        tmbchar tempvals[80] = {0};
        tmbstr vals = &tempvals[0];
        tmbstr ro   = ( isReadOnly ? "*" : "" );

        /* Handle special cases first.
        */
        switch ( optId )
        {
        case TidyIndentContent:
#if SUPPORT_UTF16_ENCODINGS
        case TidyOutputBOM:
#endif
            type = "AutoBool";
            vals = (tmbstr) tidyOptGetCurrPick( tdoc, optId );
            break;

        case TidyDuplicateAttrs:
            type = "enum";
            vals = (tmbstr) tidyOptGetCurrPick( tdoc, optId );
            break;

        case TidyDoctype:
            sval = (tmbstr) tidyOptGetCurrPick( tdoc, TidyDoctypeMode );
            type = "DocType";
            if ( !sval || *sval == '*' )
                sval = (tmbstr) tidyOptGetValue( tdoc, TidyDoctype );
            vals = (tmbstr) sval;
            break;

        case TidyCSSPrefix:
            type = "Name";
            vals = (tmbstr) tidyOptGetValue( tdoc, TidyCSSPrefix );
            break;

        case TidyInlineTags:
        case TidyBlockTags:
        case TidyEmptyTags:
        case TidyPreTags:
            {
                TidyIterator pos = tidyOptGetDeclTagList( tdoc );
                type = "Tag names";
                while ( pos )
                {
                    vals = (tmbstr) tidyOptGetNextDeclTag(tdoc, optId, &pos);
                    if ( pos )
                    {
                        if ( *name )
                            printf( valfmt, name, type, ro, vals );
                        else
                            printf( fmt, name, type, vals );
                        name = "";
                        type = "";
                    }
                }
            }
            break;

        case TidyCharEncoding:
        case TidyInCharEncoding:
        case TidyOutCharEncoding:
            type = "Encoding";
            sval = tidyOptGetEncName( tdoc, optId );
            vals = (tmbstr) sval;
            break;

        case TidyNewline:
            type = "enum";
            vals = (tmbstr) tidyOptGetCurrPick( tdoc, optId );
            break;

        /* General case will handle remaining */
        default:
            switch ( optTyp )
            {
            case TidyBoolean:
                type = "Boolean";   /* curr pick handles inverse */
                vals = (tmbstr) tidyOptGetCurrPick( tdoc, optId );
                break;

            case TidyInteger:
                type = "Integer";
                ival = tidyOptGetInt( tdoc, optId );
                sprintf( tempvals, "%d", ival );
                break;

            case TidyString:
                type = "String";
                vals = (tmbstr) tidyOptGetValue( tdoc, optId );
                break;
            }
        }

        if ( *name || *type || *vals )
        {
            if ( ! vals )
                vals = "";
            if ( *name )
                printf( valfmt, name, type, ro, vals );
            else
                printf( fmt, name, type, vals );
        }
    }

    printf( "\n\nValues marked with an *asterisk are calculated \n"
            "internally by HTML Tidy\n\n" );
}

void version( TidyDoc tdoc, ctmbstr prog )
{
#ifdef PLATFORM_NAME
    printf( "HTML Tidy for %s released on %s\n",
             PLATFORM_NAME, tidyReleaseDate() );
#else
    printf( "HTML Tidy released on %s\n", tidyReleaseDate() );
#endif
}

void unknownOption( TidyDoc tdoc, uint c )
{
    fprintf( errout, "HTML Tidy: unknown option: %c\n", c );
}

int main( int argc, char** argv )
{
    ctmbstr prog = argv[0];
    ctmbstr cfgfil = null, errfil = null, htmlfil = null;
    TidyDoc tdoc = tidyCreate();
    int status = 0;

    errout = stderr;  /* initialize to stderr */

#ifdef CONFIG_FILE
    tidyLoadConfig( tdoc, CONFIG_FILE );
#endif /* CONFIG_FILE */

    /* look for env var "HTML_TIDY" */
    /* then for ~/.tidyrc (on platforms defining $HOME) */

    if ( cfgfil = getenv("HTML_TIDY") )
        tidyLoadConfig( tdoc, cfgfil );
#ifdef SUPPORT_GETPWNAM
    else
        tidyLoadConfig( tdoc, "~/.tidyrc" );
#endif /* SUPPORT_GETPWNAM */

    /* read command line */
    while ( argc > 0 )
    {
        if (argc > 1 && argv[1][0] == '-')
        {
            /* support -foo and --foo */
            ctmbstr arg = argv[1] + 1;

            if ( strcasecmp(arg, "xml") == 0)
                tidyOptSetBool( tdoc, TidyXmlTags, yes );

            else if ( strcasecmp(arg,   "asxml") == 0 ||
                      strcasecmp(arg, "asxhtml") == 0 )
            {
                tidyOptSetBool( tdoc, TidyXhtmlOut, yes );
            }
            else if ( strcasecmp(arg,   "ashtml") == 0 )
                tidyOptSetBool( tdoc, TidyHtmlOut, yes );

            else if ( strcasecmp(arg, "indent") == 0 )
            {
                tidyOptSetInt( tdoc, TidyIndentContent, TidyYesState );
            }
            else if ( strcasecmp(arg, "omit") == 0 )
                tidyOptSetBool( tdoc, TidyHideEndTags, yes );

            else if ( strcasecmp(arg, "upper") == 0 )
                tidyOptSetBool( tdoc, TidyUpperCaseTags, yes );

            else if ( strcasecmp(arg, "clean") == 0 )
                tidyOptSetBool( tdoc, TidyMakeClean, yes );

            else if ( strcasecmp(arg, "bare") == 0 )
                tidyOptSetBool( tdoc, TidyMakeBare, yes );

            else if ( strcasecmp(arg, "raw") == 0      ||
                      strcasecmp(arg, "ascii") == 0    ||
                      strcasecmp(arg, "latin0") == 0   ||
                      strcasecmp(arg, "latin1") == 0   ||
                      strcasecmp(arg, "utf8") == 0     ||
                      strcasecmp(arg, "iso2022") == 0  ||
#if SUPPORT_UTF16_ENCODINGS
                      strcasecmp(arg, "utf16le") == 0  ||
                      strcasecmp(arg, "utf16be") == 0  ||
                      strcasecmp(arg, "utf16") == 0    ||
#endif
#if SUPPORT_ASIAN_ENCODINGS
                      strcasecmp(arg, "shiftjis") == 0 ||
                      strcasecmp(arg, "big5") == 0     ||
#endif
                      strcasecmp(arg, "mac") == 0      ||
                      strcasecmp(arg, "win1252") == 0  ||
                      strcasecmp(arg, "ibm858") == 0 )
            {
                tidySetCharEncoding( tdoc, arg );
            }
            else if ( strcasecmp(arg, "numeric") == 0 )
                tidyOptSetBool( tdoc, TidyNumEntities, yes );

            else if ( strcasecmp(arg, "modify") == 0 ||
                      strcasecmp(arg, "change") == 0 ||  /* obsolete */
                      strcasecmp(arg, "update") == 0 )   /* obsolete */
            {
                tidyOptSetBool( tdoc, TidyWriteBack, yes );
            }
            else if ( strcasecmp(arg, "errors") == 0 )
                tidyOptSetBool( tdoc, TidyShowMarkup, no );

            else if ( strcasecmp(arg, "quiet") == 0 )
                tidyOptSetBool( tdoc, TidyQuiet, yes );

            else if ( strcasecmp(arg, "help") == 0 ||
                      strcasecmp(arg,    "h") == 0 || *arg == '?' )
            {
                help( tdoc, prog );
                tidyRelease( tdoc );
                return 0; /* success */
            }
            else if ( strcasecmp(arg, "help-config") == 0 )
            {
                optionhelp( tdoc, prog );
                tidyRelease( tdoc );
                return 0; /* success */
            }
            else if ( strcasecmp(arg, "show-config") == 0 )
            {
                optionvalues( tdoc, prog );
                tidyRelease( tdoc );
                return 0; /* success */
            }
            else if ( strcasecmp(arg, "config") == 0 )
            {
                if ( argc >= 3 )
                {
                    ctmbstr post;

                    tidyLoadConfig( tdoc, argv[2] );

                    /* Set new error output stream if setting changed */
                    post = tidyOptGetValue( tdoc, TidyErrFile );
                    if ( post && (!errfil || !samefile(errfil, post)) )
                    {
                        errfil = post;
                        errout = tidySetErrorFile( tdoc, post );
                    }

                    --argc;
                    ++argv;
                }
            }

#if SUPPORT_ASIAN_ENCODINGS
            else if ( strcasecmp(arg, "language") == 0 ||
                      strcasecmp(arg,     "lang") == 0 )
            {
                if ( argc >= 3 )
                {
                    tidyOptSetValue( tdoc, TidyLanguage, argv[2] );
                    --argc;
                    ++argv;
                }
            }
#endif

            else if ( strcasecmp(arg, "output") == 0 ||
                      strcasecmp(arg, "-output-file") == 0 ||
                      strcasecmp(arg, "o") == 0 )
            {
                if ( argc >= 3 )
                {
                    tidyOptSetValue( tdoc, TidyOutFile, argv[2] );
                    --argc;
                    ++argv;
                }
            }
            else if ( strcasecmp(arg,  "file") == 0 ||
                      strcasecmp(arg, "-file") == 0 ||
                      strcasecmp(arg,     "f") == 0 )
            {
                if ( argc >= 3 )
                {
                    errfil = argv[2];
                    errout = tidySetErrorFile( tdoc, errfil );
                    --argc;
                    ++argv;
                }
            }
            else if ( strcasecmp(arg,  "wrap") == 0 ||
                      strcasecmp(arg, "-wrap") == 0 ||
                      strcasecmp(arg,     "w") == 0 )
            {
                if ( argc >= 3 )
                {
                    uint wraplen = 0;
                    sscanf( argv[2], "%d", &wraplen );
                    tidyOptSetInt( tdoc, TidyWrapLen, wraplen );
                    --argc;
                    ++argv;
                }
            }
            else if ( strcasecmp(arg,  "version") == 0 ||
                      strcasecmp(arg, "-version") == 0 ||
                      strcasecmp(arg,        "v") == 0 )
            {
                version( tdoc, prog );
                tidyRelease( tdoc );
                return 0;  /* success */

            }
            else if ( strncmp(argv[1], "--", 2 ) == 0)
            {
                if ( tidyOptParseValue(tdoc, argv[1]+2, argv[2]) )
                {
                    ++argv;
                    --argc;
                }
            }

#if SUPPORT_ACCESSIBILITY_CHECKS
            else if ( strcasecmp(arg, "access") == 0 )
            {
                if ( argc >= 3 )
                {
                    uint acclvl = 0;
                    sscanf( argv[2], "%d", &acclvl );
                    tidyOptSetInt( tdoc, TidyAccessibilityCheckLevel, acclvl );
                    --argc;
                    ++argv;
                }
            }
#endif

            else
            {
                uint c;
                ctmbstr s = argv[1];

                while ( c = *++s )
                {
                    switch ( c )
                    {
                    case 'i':
                        tidyOptSetInt(tdoc, TidyIndentContent, TidyYesState);
                        break;

                    /* Usurp -o for output file.  Anyone hiding end tags?
                    case 'o':
                        tidyOptSetBool( tdoc, TidyHideEndTags, yes );
                        break;
                    */

                    case 'u':
                        tidyOptSetBool( tdoc, TidyUpperCaseTags, yes );
                        break;

                    case 'c':
                        tidyOptSetBool( tdoc, TidyMakeClean, yes );
                        break;

                    case 'b':
                        tidyOptSetBool( tdoc, TidyMakeBare, yes );
                        break;

                    case 'n':
                        tidyOptSetBool( tdoc, TidyNumEntities, yes );
                        break;

                    case 'm':
                        tidyOptSetBool( tdoc, TidyWriteBack, yes );
                        break;

                    case 'e':
                        tidyOptSetBool( tdoc, TidyShowMarkup, no );
                        break;

                    case 'q':
                        tidyOptSetBool( tdoc, TidyQuiet, yes );
                        break;

                    default:
                        unknownOption( tdoc, c );
                        break;
                    }
                }
            }

            --argc;
            ++argv;
            continue;
        }

        if ( argc > 1 )
        {
            htmlfil = argv[1];
            if ( tidyOptGetBool(tdoc, TidyEmacs) )
                tidyOptSetValue( tdoc, TidyEmacsFile, htmlfil );
            status = tidyParseFile( tdoc, htmlfil );
        }
        else
        {
            htmlfil = "stdin";
            status = tidyParseStdin( tdoc );
        }

        if ( status >= 0 )
            status = tidyCleanAndRepair( tdoc );

        if ( status >= 0 )
            status = tidyRunDiagnostics( tdoc );

        if ( status >= 0 )
        {
            if ( tidyOptGetBool(tdoc, TidyWriteBack) )
                status = tidySaveFile( tdoc, htmlfil );
            else
            {
                ctmbstr outfil = tidyOptGetValue( tdoc, TidyOutFile );
                if ( outfil )
                    status = tidySaveFile( tdoc, outfil );
                else
                    status = tidySaveStdout( tdoc );
            }
        }

        contentErrors   += tidyErrorCount( tdoc );
        contentWarnings += tidyWarningCount( tdoc );
        accessWarnings  += tidyAccessWarningCount( tdoc );

        --argc;
        ++argv;

        if ( argc <= 1 )
            break;
    }

    if ( contentErrors + contentWarnings > 0 && 
         !tidyOptGetBool(tdoc, TidyQuiet) )
    {
        tidyErrorSummary( tdoc );
        tidyGeneralInfo( tdoc );
    }

    /* called to free hash tables etc. */
    tidyRelease( tdoc );

    /* return status can be used by scripts */
    if ( contentErrors > 0 )
        return 2;

    if ( contentWarnings > 0 )
        return 1;

    /* 0 signifies all is ok */
    return 0;
}

